csharp-mode.el

```diff
@@ -2077,1076 +1837,282 @@ to the beginning of the prior namespace.
+             (container-p1 (car (split-string (car container))))
+             ;; class/interface
+             (container-p2 (cadr (split-string (car container)))))
+        ;; use p1 (namespace) when there is no p2
+        (if container-p2
+            container-p2
+          container-p1)))))
+
+(defun csharp--imenu-sort (items)
+  (sort items (lambda (item1 item2)
+                (string< (car item1) (car item2)))))
+
+(defun csharp--imenu-get-class-name (class namespaces)
+  (let ((namespace (csharp--imenu-get-container-name class namespaces))
+        (class-name (car class)))
+    (if (not namespace)
+        class-name
+      ;; reformat to include namespace
+      (let* ((words (split-string class-name))
+             (type  (car words))
+             (name  (cadr words)))
+        (concat type " " namespace "." name)))))
+
+(defun csharp--imenu-get-class-nodes (classes namespaces)
+  "Creates a new alist with classes as root nodes with namespaces added."
+
+  (mapcar (lambda (class)
+            (let* ((class-name (csharp--imenu-get-class-name class namespaces))
+                   (class-pos  (cdr class)))
+              (cons class-name
+                    (list
+                     (cons "( top )" class-pos)))))
+          classes))
+
+(defun csharp--imenu-get-class-node (result item classes namespaces)
+  (let* ((class-item (csharp--imenu-get-container item classes nil))
+         (class-name (csharp--imenu-get-class-name class-item namespaces)))
+    (assoc class-name result)))
+
+(defun csharp--imenu-format-item-node (item type)
+  (cons
+   (concat "(" type ") " (car item))
+   (cdr item)))
+
+(defun csharp--imenu-append-items-to-menu (result type index classes namespaces)
+  ;; items = all methods, all events, etc based on "type"
+  (let* ((items (cdr (assoc type index))))
+    (dolist (item items)
+      (let ((class-node (csharp--imenu-get-class-node result item classes namespaces))
+            (item-node  (csharp--imenu-format-item-node item type)))
+        (nconc class-node (list item-node))))))
+
+(defun csharp--imenu-transform-index (index)
+  (let* ((result nil)
+         (namespaces (cdr (assoc "namespace" index)))
+         (classes    (cdr (assoc "class"     index)))
+         (class-nodes (csharp--imenu-get-class-nodes classes namespaces)))
+    ;; be explicit about collection variable
+    (setq result class-nodes)
+    (dolist (type '("ctor" "method" "method-inf" "prop" "field" "event" "indexer"))
+      (csharp--imenu-append-items-to-menu result type index classes namespaces))
+
+    ;; add enums to main result list, as own items.
+    ;; We don't support nested types. EOS.
+    (dolist (enum (cdr (assoc "enum" index)))
+      (let ((enum-name (csharp--imenu-get-class-name enum namespaces)))
+        (setq result (cons (cons enum-name (cdr enum)) result))))
+
+    ;; sort individual sub-lists
+    (dolist (item result)
+      (when (listp (cdr item))
+        (setcdr item (csharp--imenu-sort (cdr item)))))
```

Comment:
Thanks. That's what I was looking for, bit didn't bother looking up in my eagerness to get this code working :)